src/igraph_eulerian.c

```diff
@@ -0,0 +1,499 @@
+}
+
+
+int igraph_eulerian_path_directed(igraph_t *graph, igraph_vector_t *res) {
+    igraph_bool_t res_strong, res_weak;
+    igraph_integer_t incoming_excess, outgoing_excess;
+    igraph_integer_t start_node;
+    igraph_t copy;
+    igraph_integer_t i;
+    igraph_vector_t out_degree, in_degree;
+
+    IGRAPH_CHECK(igraph_copy(&copy, graph));
+    IGRAPH_FINALLY(igraph_destroy, &copy);
+
+    incoming_excess = 0;
+    outgoing_excess = 0;
+
+    /* determining the start node */
+    /* also getting the outgoing list for each vector*/
+
+    IGRAPH_CHECK(igraph_vector_init(&out_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &out_degree);
+    igraph_degree(graph, &out_degree, igraph_vss_all(), IGRAPH_OUT, IGRAPH_LOOPS);
+
+    IGRAPH_CHECK(igraph_vector_init(&in_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &in_degree);
+    igraph_degree(graph, &in_degree, igraph_vss_all(), IGRAPH_IN, IGRAPH_LOOPS);
+
+    /* checking if incoming vertices == outgoing vertices */
+    for (i = 0; i < igraph_vcount(graph); i++) {
+        if (VECTOR(in_degree)[i] != VECTOR(out_degree)[i]) {
+            if ((VECTOR(in_degree)[i] + 1 == VECTOR(out_degree)[i]) && (incoming_excess < 2 && outgoing_excess < 1)) {
+                outgoing_excess++;
+                start_node = i;
+            } else if ((VECTOR(out_degree)[i] + 1 == VECTOR(in_degree)[i]) && (incoming_excess < 1 && outgoing_excess < 2)) {
+                incoming_excess++;
+            } 
+        }
+    }
+
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_strong, IGRAPH_STRONG));
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_weak, IGRAPH_WEAK));
+
+    if ((outgoing_excess == 0 && incoming_excess == 0) && (res_strong)) {
+        start_node = 0;
+        eulerian_path_directed_implementation(&copy, &start_node, &out_degree, res);
+    } else if ((outgoing_excess == 1 && incoming_excess == 1) && 
+        (res_strong || res_weak)) {
+        eulerian_path_directed_implementation(&copy, &start_node, &out_degree, res);
+    }
+
+    igraph_vector_destroy(&in_degree);
+    igraph_vector_destroy(&out_degree);
+    igraph_destroy(&copy);
+    IGRAPH_FINALLY_CLEAN(3);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+int igraph_eulerian_cycle(igraph_t *graph, igraph_vector_t *res) {
+    igraph_bool_t cycle;
+    igraph_bool_t has_path;
+
+    has_path = 0;
+    cycle = 0;
+
+    igraph_is_eulerian(graph, &has_path, &cycle);
+
+    if (!cycle) {
+        IGRAPH_WARNING("Eulerian cycle not possible");
+        return IGRAPH_FAILURE;
+    }
+
+    if (igraph_is_directed(graph)) {
+        igraph_eulerian_path_directed(graph, res);
+    } else {
+        igraph_euler_path_undirected(graph, res);
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+int igraph_eulerian_path(igraph_t *graph, igraph_vector_t *res) {
+    igraph_bool_t cycle;
+    igraph_bool_t has_path;
+
+    has_path = 0;
+    cycle = 0;
+
+    igraph_is_eulerian(graph, &has_path, &cycle);
+
+    if (!has_path && !cycle) {
```

Comment:
Why is it necessary to check `&& ! cycle`?

Reply:
Good comment. If a eulerian cycle exists, then that also means a path exists. I have decided to have that function (eulerian_path) return an eulerian cycle (if it exists).

Reply:
When I was writing this, I had my is_eulerian function have has_path equal to zero if a cycle exists. That is why I had that check. I feel however that my implementation may not make sense would you agreee?

Reply:
Let's change it such that when cycle is 1, has_path is also 1.